openai_llm.py

```diff
@@ -0,0 +1,17 @@
+from langchain_openai import ChatOpenAI
+import os
+import time
+from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type
+import openai  # Ensure the openai library is correctly imported
+
+class OpenAIWrapper:
+    def __init__(self, api_key):
+        self.api_key = api_key
+        os.environ["OPENAI_API_KEY"] = self.api_key
+        os.environ["OPENAI_MODEL_NAME"] = "gpt-3.5-turbo"  # Update the model name here
+        self.llm = ChatOpenAI(model="gpt-3.5-turbo")
+
+    @retry(stop=stop_after_attempt(5), wait=wait_exponential(multiplier=1, max=60), retry=retry_if_exception_type(openai.RateLimitError))
```

Comment:
When I run the program now I get the following rate limit error `Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}`. So I am rate limited because I don't have any OpenAI credits. So the rate limit is not that I sent too many requests too fast, it is that I sent requests at all.

Is it worth separating out these issues, or just let it retry anyway? I can't think of anything it would hurt to let it retry.

Reply:
Best handle this specific case separately rather than retrying. Retrying will only waste time and resources without resolving the underlying issue. 

We can handle the  insufficient_quota  error separately

- **Error Handling**: 
  - The  generate  method now includes a  try-except  block for  openai.error.RateLimitError . 
  - If the error code is  insufficient_quota , it raises a  RuntimeError  to avoid unnecessary retries. 
  - Other  RateLimitError  types are handled by the retry mechanism as before.

Reply:
`crew.kickoff` seems to do something with openai because I was getting the error about insufficient_quota here and it never made it to the generate. I pushed an update to changed where it does the retry and that fixed the issue. see what you think.

Reply:
That looks good. Glad we got that fixed!